Add FFI Functions for Clamp Control

MichaelHuth opened the pull request:
close #2547

MichaelHuth committed:
DAP: Refactor, Moved setting of clamp mode to own DAP function

this was previously done directly in the GUI event handler

MichaelHuth committed:
FFI: Add functions for clamp and headstage control

Added the following functions:

FFI_GetCurrentClampState
returns a wave with the current clamp control values

FFI_GetClampState
returns a wave with the all clamp control values

FFI_SetClampMode

FFI_SetHoldingPotential
- allowed only in VC mode

FFI_SetBiasCurrent
- allowed only in IC mode

FFI_SetAutoBias
- allowed only in IC mode

FFI_SetHeadstageActive
- allowed only when no DAQ is running

FFI_GetHeadstageActive
- returns 1 is the given headstage is active, 0 otherwise

FFI_TriggerAutoClampControl
- wraps Pipette, Capacitance and Bridge Balance auto setting in a single call

MichaelHuth committed:
AI: Add range check for MCC_NO_AUTOBIAS_V_FUNC value write

In MIES the range is limited through the corresponding GUi element.
However, when changed externally through a ZeroMQ function call then
the range was not checked.

t-b commented:
Let's do your usual appraoch here where Luke/Tim can test the FFI functions before we write the tests.

MichaelHuth commented:
If the current clamp mode for the headstage in MIES is different from the specific calmp mode of an amplifier setting (e.g. auto bridge balance that applies only to IC) then it is ignored by MIES.

campagnola commented:
> If the current clamp mode for the headstage in MIES is different from the specific calmp mode of an amplifier setting (e.g. auto bridge balance that applies only to IC) then it is ignored by MIES.

That makes sense for any of the TriggerAuto* functions, but all other parameters I would expect to be settable even when the mode is not active (for example, I should be able to set the auto bias parameters before switching to IC)

MichaelHuth committed:
AI: Refactor AI_AmpStorageControlToRowLabel for correct VC/IC block

With AI_GetFunctionConstantForClampMode the relevant function numbers
can be retrieved and with AI_MapFunctionConstantToControl the
corresponding GUI control names.
This gives a list of GUI controls for each clamp mode.
In AI_AmpStorageControlToRowLabel the block was incorrectly sorted in
the case chain.

MichaelHuth commented on Packages/MIES/MIES_ForeignFunctionInterface.ipf:
```diff
@@ -574,3 +582,257 @@ Function RunTests()
+		return clampState
+	endif
+
+	WAVE aiFuncs = AI_GetFunctionConstantForClampMode(clampMode)
+	numFuncs = DimSize(aiFuncs, ROWS)
+	Make/FREE/T/N=(numFuncs) ctrlNames, ampLabels
+
+	ctrlNames[] = AI_MapFunctionConstantToControl(aiFuncs[p], clampMode)
+	ampLabels[] = AI_AmpStorageControlToRowLabel(ctrlNames[p])
+
+	Make/FREE/D/N=(numFuncs + 1) clampState
+	for(i = 0; i < numFuncs; i += 1)
+		if(IsEmpty(ampLabels[i]))
+			continue
+		endif
+		clampState[j] = AmpStorageWave[%$ampLabels[i]][0][headStage]
+		lbl           = AI_MapFunctionConstantToName(aiFuncs[i], clampMode)
+		SetDimLabel ROWS, j, $lbl, clampState
+		j += 1
+	endfor
```
Gets fixed with #2704

MichaelHuth commented:
@campagnola I changed the following:

- previous `FFI_GetClampState` is now `FFI_GetCurrentClampState`
- New `FFI_GetClampState` returns now the Amplifier Storage Wave from MIES for the specified headstage with all amplifier settings. The wave has dimension labels and the function has docu.
- `FFI_SetHoldingPotential` and `FFI_SetBiasCurrent` have an additional enable argument that can be 0 (disable) and 1 (enable). When you set enable the value is written first then it is enabled, if you disable then it is first disabled and then the value set.
- `FFI_SetAutoBias` uses the same order for value setting and enabling as `FFI_SetHoldingPotential` and `FFI_SetBiasCurrent`
- New `FFI_GetHeadstageActive` returns 1 or 0, depending if the given headstage is enabled or not.
- `FFI_SetHeadstageActive` raises an assertion when attempting to activate a headstage while data acquisition is running
- For `FFI_SetClampMode` the clamp mode change is deferred if data acquisition is running until it stopped. If that is problematic I can also add an assertion there for this case.
